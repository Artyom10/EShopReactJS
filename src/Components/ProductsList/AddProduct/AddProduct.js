import React, { useState } from 'react';
import stylesFor from './AddProduct.module.css';
import AddProductButton from '../AddPoductButton/AddProductButton';


const AddProduct = (props) => {

  const [newProduct, setNewProduct] = useState({
    photo: '',
    price: '',
    producer: '',
    type: '',
    size: '',
    tags: '',
    description: '',
    rated: [],
  });

  const handleChange = (e) => {
    e.preventDefault();
    setNewProduct({
      ...newProduct,
      [e.target.id]: e.target.value
    });
  };

  

 const handleSubmit = (e) => {
    e.preventDefault();
    props.addNewProduct(newProduct);
    setNewProduct({
      photo: '',
      producer: '',
      price: '',
      type: '',
      size: '',
      tags: '',
      description: '',
  });
};

    return (
      <div className="container">
      <div className="row  justify-content-center">
        <div className="col-md-5">
      <div className="align-items-center">
       <div className={`${stylesFor.loginContainer} d-flex  align-items-center justify-content-center`}>
         <form onSubmit={handleSubmit} className={`${stylesFor.loginForm} text-center`}>
           <h1 className="mb-5 font-weight-light text-uppercase">Add new product</h1>
           <div className="form-group">
            <input type="text" className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
            placeholder="photo" onChange={handleChange} id="photo" value={newProduct.photo} />
          </div>
          <div className="form-group">
            <input type="text" className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
            placeholder="producer" onChange={handleChange} id="producer" value={newProduct.producer} />
          </div>
          <div className="form-group">
            <input type="text" className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
            placeholder="type" onChange={handleChange} id="type" value={newProduct.type} />
          </div>
          <div className="form-group">
            <input type="text" className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
            placeholder="price" onChange={handleChange} id="price" value={newProduct.price} />
          </div>
          <div className="form-group">
            <input type="text" className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
            placeholder="size" onChange={handleChange} id="size" value={newProduct.size}/>
          </div>
          <div className="form-group">
            <input type="text" className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
            placeholder="tags" onChange={handleChange} id="tags" value={newProduct.tags} />
          </div>
          <div className="form-group">
            <input type="text" className={`form-control rounded-pill form-control-lg ${stylesFor.control}`}
            placeholder="description" onChange={handleChange} id="description" value={newProduct.description} />
          </div>
           <AddProductButton />
        </form>
       </div>
      </div>
      </div>
    </div>
  </div>
    );
}

export default AddProduct;