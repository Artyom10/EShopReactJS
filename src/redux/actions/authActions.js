export const signIn = (credentials) => {
    debugger;
    return (dispatch, getState, {getFirebase}) => {
        const firebase = getFirebase();

        firebase.auth().signInWithEmailAndPassword(
            credentials.email,
            credentials.password
        ).then(() => {
            dispatch({type: 'LOGIN_SUCCESS'});
        }).catch((err) => {
            dispatch({type: 'LOGIN_ERROR', err})
        });
    }
};

export const signOut = () => {
    return(dispatch, getState, {getFirebase}) => {
        const firebase = getFirebase();

        firebase.auth().signOut().then(() => {
            dispatch({type: 'SIGNOUT_SUCCESS'})
        });
    } 
}

export const signUp = (newUser) => {
    return(dispatch, getState, {getFirebase, getFirestore}) => {
        debugger;
        const firebase = getFirebase();
        const firestore = getFirestore();

        firebase.auth().createUserWithEmailAndPassword(
            newUser.email,
            newUser.password,
        ).then((response) => {
            return firestore.collection('users').doc(response.user.uid).set({
                firstName: newUser.firstName,
                secondName: newUser.secondName,
                request: false, //newUser.request,
                bags: [], //newUser.bags,
                userPhoto: 'https://icons-for-free.com/iconfiles/png/512/business+costume+male+man+office+user+icon-1320196264882354682.png',
                valuedProducts: [],//newUser.valuedProducts,
                isAdmin: false,
                isWho: 'user',
                email: newUser.email,
            })
        }).then(() => {
            dispatch({type: 'SIGNUP_SUCCESS'})
        }).catch(err => {
            dispatch({type: 'SIGNUP_ERROR', err})
        })
    }
}