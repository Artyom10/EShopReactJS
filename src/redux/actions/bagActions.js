/*export const removeFromBag = (removeTargetBag) => {
    return (dispatch, getState) => {
 dispatch({ type: 'REMOVE_FROM_BAG', removeTargetBag,})
    }
}*/

export const addToBag = (targetAddToBag) => {
    return (dispatch, getState,{getFirebase, getFirestore}) => {
     const firestore = getFirestore();
     //const profile = getState()
     const buyerId = getState().firebase.auth.uid
     firestore.collection('users').doc(buyerId).update({
            bags: [targetAddToBag]
     })
     .then(() => {
         dispatch({type: 'ADD_TO_BAG', targetAddToBag})
     }).catch((err) => {
         dispatch({type: 'ADD_TO_BAG_ERROR', err})
     })
    }
}


 //dispatch({type: 'ADD_TO_BAG', targetAddToBag})