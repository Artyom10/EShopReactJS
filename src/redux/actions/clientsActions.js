import API from '../../services/axios';

export const deleteUser = (uid) => {
    return(dispatch,getState,{getfirebase, getFirestore}) => {
        API.post('clients', uid);
    }

}