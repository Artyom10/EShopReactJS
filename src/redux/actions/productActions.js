export const addNewProduct = (newProduct) => {
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    const firestore = getFirestore();
    const storage = getFirebase().storage();

    const infoPhoto = newProduct.photo;
    newProduct.photo = null;

    firestore
      .collection("products")
      .add({
        ...newProduct,
        isBought: false,
      })
      .then((response) => {
        infoPhoto &&
        storage
          .ref(`photos/${response.id}`)
          .put(infoPhoto)
          .on(
            'state_changed',
            snapshot => { },
            error => { console.log(error) },
            () => {
              storage
                .ref('photos')
                .child(response.id)
                .getDownloadURL()
                .then(url => {
                  newProduct.photo = url
                  return firestore
                    .collection('products')
                    .doc(response.id)
                    .update({photo: newProduct.photo})
                })
            }
          )
      })
      .then(() => {
        dispatch({ type: "ADD_PRODUCT", newProduct });
      })
      .catch((err) => {
        dispatch({ type: "ADD_PRODUCT_ERROR", err });
      });
  };
};

export const removeProduct = (targetDeleteProduct) => {
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    console.log(getState());
    const firestore = getFirestore();
    const storage = getFirebase().storage();
    firestore
      .collection("products")
      .doc(targetDeleteProduct)
      .get()
      .then((doc) => {
        firestore
          .collection('products')
          .doc(targetDeleteProduct)
          .delete()
          .then(() => {
            doc.data().photo && 
            storage
              .ref(`photos/${targetDeleteProduct}`)
              .delete()
          })
      })
      .then(() => {
        dispatch({ type: "REMOVE_PRODUCT", targetDeleteProduct });
      })
      .catch((err) => {
        dispatch({ type: "REMOVE_PRODUCT_ERROR", err });
      });

      const buyerId = getState().firebase.auth.uid;
      const firebaseBuyer = getState().firebase;
      let findIndex = null;

      let ratedProducts = firebaseBuyer.profile.valuedProducts;
      ratedProducts.forEach((product,index) => {
        if(product.targetProductRating === targetDeleteProduct){
                findIndex = index;       
        }
      });
      ratedProducts.splice(findIndex,1);

      let findBagsIndex = null;

      let bagsProducts = firebaseBuyer.profile.bags;
      bagsProducts.forEach((product,index) => {
        if(product === targetDeleteProduct){
          findBagsIndex = index
        }
      });
      bagsProducts.splice(findBagsIndex,1);
  
      firestore.collection('users').doc(buyerId).update({
       valuedProducts:   ratedProducts && [...ratedProducts],
       bags: bagsProducts && [...bagsProducts],
      })
      .then(() => {
        dispatch({ type: "REMOVE_PRODUCT" });
      })
      .catch((err) => {
        dispatch({ type: "REMOVE_PRODUCT_ERROR", err });
      });


  };
};

export const editProduct = (targetEditProduct, changes) => {
  debugger;
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    const firestore = getFirestore();
    const storage = getFirebase().storage();
    typeof changes.photo !== 'string' &&  changes.photo
    ?
    storage
      .ref('photos')
      .child(targetEditProduct)
      .put(changes.photo)
      .on(
        'state_changed',
        snapshot => { },
        error => { console.log(error) },
        () => {
          storage
           .ref('photos')
           .child(targetEditProduct)
           .getDownloadURL()
           .then(url => {
             changes.photo = url
             return firestore
               .collection('products')
               .doc(targetEditProduct)
               .update({
                 ...changes
               })
           })
           .then(() => {
            dispatch({ type: "EDIT_PRODUCT", targetEditProduct });
          })
          .catch((err) => {
            dispatch({ type: "EDIT_PRODUCT_ERROR", err });
          });
        }
      )
      :
      firestore
        .collection('products')
        .doc(targetEditProduct)
        .update({
          ...changes
        })
        .then(() => {
          dispatch({ type: "EDIT_PRODUCT", targetEditProduct });
        })
        .catch((err) => {
          dispatch({ type: "EDIT_PRODUCT_ERROR", err });
        });
  };
};

export const buyProduct = (targetProductBuy) => {
  return (dispatch, getState, { getFirebase, getFirestore }) => {
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
    firestore
      .collection("products")
      .doc(targetProductBuy)
      .update({
        isBought: true,
      })
      .then(() => {
        dispatch({ type: "UPDATE_BOUGHT_PRODUCT_STATUS", targetProductBuy });
      })
      .catch((err) => {
        dispatch({ type: "AUPDATE_BOUGHT_PRODUCT_STATUS_ERROR", err });
      });

    firestore
      .collection("users")
      .doc(buyerId)
      .update({
        bags: firebaseBuyer.profile.bags
          ? [...firebaseBuyer.profile.bags, targetProductBuy]
          : [targetProductBuy],
      })
      .then(() => {
        dispatch({ type: "ADD_TO_BOUGHT", targetProductBuy });
      })
      .catch((err) => {
        dispatch({ type: "ADD_TO_BOUGHT_ERROR", err });
      });
  };
};

export const setRating = (targetProductRating, value, certainProduct) => {
  return (dispatch, getState, {getFirebase, getFirestore}) => {
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
  
    if(firebaseBuyer.profile.valuedProducts.some((everyData) => everyData.targetProductRating === targetProductRating)){
    console.log('You have already rated this product, please delete your rate and try again');
    }
    else{
          console.log('This product does not rate yet, we need update product and value')
          firestore.collection('users').doc(buyerId).update({
            valuedProducts: [...firebaseBuyer.profile.valuedProducts, {targetProductRating, value,}],
        })
        .then(() => {
          dispatch({ type: "RATE_PRODUCT", targetProductRating });
        })
        .catch((err) => {
          dispatch({ type: "RATE_PRODUCT_ERROR", err });
        });
    }
  }
}

export const deleteRating = (targetProductDeleteRating) => {
  debugger;
  return (dispatch, getState, {getFirebase, getFirestore}) => {
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
    let findIndex = null;

    let ratedProducts = firebaseBuyer.profile.valuedProducts;
    ratedProducts.forEach((product,index) => {
      if(product.targetProductRating === targetProductDeleteRating){
              findIndex = index;       
      }
    });
    ratedProducts.splice(findIndex,1);

    firestore.collection('users').doc(buyerId).update({
     valuedProducts:   ratedProducts && [...ratedProducts]
    })
    .then(() => {
      dispatch({ type: "DELETE_RATE_PRODUCT", targetProductDeleteRating });
    })
    .catch((err) => {
      dispatch({ type: "DELETE_RATE_PRODUCT_ERROR", err });
    });
  }
}


export const deleteBooked = (productId) => {
  return (dispatch, getState, {getFirebase, getFirestore}) => {
    debugger;
    const firestore = getFirestore();
    const buyerId = getState().firebase.auth.uid;
    const firebaseBuyer = getState().firebase;
    let findIndex = null;

    let bagsProducts = firebaseBuyer.profile.bags;
    bagsProducts.forEach((product,index) => {
      if(product === productId){
              findIndex = index;       
      }
    });
    bagsProducts.splice(findIndex,1);

    firestore.collection('products').doc(productId).update({
      isBought: false,
    })
    .catch((err) => {
      dispatch({type: "DELETE_BAG_PRODUCT_ERROR", err});
    })
    
    firestore.collection('users').doc(buyerId).update({
      bags:   bagsProducts && [...bagsProducts]
     })
     .then(() => {
       dispatch({ type: "DELETE_BAG_PRODUCT" });
     })
     .catch((err) => {
       dispatch({ type: "DELETE_BAG_PRODUCT_ERROR", err });
     });
   }
}