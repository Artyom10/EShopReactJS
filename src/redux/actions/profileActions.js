export const updateProfile = (firstName, secondName) => {
    return (dispatch, getState,{getFirebase, getFirestore}) => {
     const firestore = getFirestore();
     const profile = getState().firebase.auth.uid
     firestore.collection('users').doc(profile).update({
            firstName,
            secondName,
     })
     .then(() => {
         dispatch({type: 'UPDATE_PROFILE'})
     }).catch((err) => {
         dispatch({type: 'UPDATE_PROFILE_ERROR', err})
     })
    }
}

export const removeRequest = (user) => {
    debugger;
    return (dispatch, getState, {getFirebase, getFirestore}) => {
        const firestore = getFirestore();
        const profile = getState().firebase.auth.uid
        firestore.collection('users').doc(profile).update({
            request: true,
        })
        .then(() => {
            dispatch({type: 'UPDATE_PROFILE_REQUEST'})
        }).catch((err) => {
            dispatch({type: 'UPDATE_PROFILE_REQUEST_ERROR', err})
        })
    }
}