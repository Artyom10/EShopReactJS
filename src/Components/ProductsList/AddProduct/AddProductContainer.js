import React from 'react';
import {addNewProduct} from '../../../redux/actions/productActions';
import {connect} from 'react-redux';
import AddProduct from './AddProduct';

const mapStateToProps = (state) => {
 return {
   
 }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addNewProduct: (newProduct) => dispatch(addNewProduct(newProduct))
  }
}

export default connect(mapStateToProps,mapDispatchToProps)(AddProduct);