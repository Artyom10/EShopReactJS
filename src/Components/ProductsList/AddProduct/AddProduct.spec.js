import { shallow } from 'enzyme';
import React from 'react';
import AddProduct from './AddProduct';
import { addNewProduct } from '../../../redux/actions/productActions';
import { render } from '@testing-library/react';

const props = {
    addNewProduct,
}

const setUp = (props) => shallow(<AddProduct {...props} />)

describe('AddProduct component', () => {
    it('should render AddProduct component', () => {
        const component = setUp(props);
        expect(component.debug()).toMatchSnapshot();
    });
});