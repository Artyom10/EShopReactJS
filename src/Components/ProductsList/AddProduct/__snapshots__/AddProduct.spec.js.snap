// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProduct component should render AddProduct component 1`] = `
"<div className=\\"container\\">
  <div className=\\"row  justify-content-center\\">
    <div className=\\"col-md-5\\">
      <div className=\\"align-items-center\\">
        <div className=\\"loginContainer d-flex  align-items-center justify-content-center\\">
          <form onSubmit={[Function: handleSubmit]} className=\\"loginForm text-center\\">
            <h1 className=\\"mb-5 font-weight-light text-uppercase\\">
              Add new product
            </h1>
            <div className=\\"form-group\\">
              <label className=\\"btn\\">
                Photo
                <input type=\\"file\\" className=\\"setPhoto\\" id=\\"photo\\" onChange={[Function: handleChange]} />
              </label>
            </div>
            <div className=\\"form-group\\">
              <input type=\\"text\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"producer\\" onChange={[Function: handleChange]} id=\\"producer\\" value=\\"\\" />
            </div>
            <div className=\\"form-group\\">
              <input type=\\"text\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"type\\" onChange={[Function: handleChange]} id=\\"type\\" value=\\"\\" />
            </div>
            <div className=\\"form-group\\">
              <input type=\\"text\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"price\\" onChange={[Function: handleChange]} id=\\"price\\" value=\\"\\" />
            </div>
            <div className=\\"form-group\\">
              <input type=\\"text\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"size\\" onChange={[Function: handleChange]} id=\\"size\\" value=\\"\\" />
            </div>
            <div className=\\"form-group\\">
              <input type=\\"text\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"tags\\" onChange={[Function: handleChange]} id=\\"tags\\" value=\\"\\" />
            </div>
            <div className=\\"form-group\\">
              <input type=\\"text\\" className=\\"form-control rounded-pill form-control-lg control\\" placeholder=\\"description\\" onChange={[Function: handleChange]} id=\\"description\\" value=\\"\\" />
            </div>
            <AddProductButton />
          </form>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
